File: tests/read_at.c
/* Actual cfg cases, address translation, diagnostics and SD read implementation. */
#include <assert.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include "sd.h"
#include "read-at-types.inc"
#define READ_AT_MAX_BLOCKS (DATA_BUFFER_SIZE / SD_SECTOR_SIZE)
_Static_assert(DATA_BUFFER_SIZE == sizeof(((sc64_buffers_t *)0)->BUFFER), "controller/N64 BRAM capacity mismatch");
_Static_assert(READ_AT_MAX_BLOCKS == 16, "READ_AT capability block limit changed");
static struct { uint32_t data[2],sd_card_sector; } p;
static struct {bool card_initialized,byte_swap,card_type_block;} sdstate;
static unsigned lock_fail,fault,lock_calls,on,off,starts,aborts,cmd18,cmd12,adc_calls;
static uint32_t reply0,reply1,status,read_address,read_count;
static unsigned char bram[DATA_BUFFER_SIZE];
static void cfg_cmd_reply_success(void){reply0=p.data[0];reply1=p.data[1];status=0;}
static void cfg_cmd_reply_error(unsigned type,uint32_t error){reply0=type<<24|error;reply1=0;status=1;}
sd_error_t sd_get_lock(sd_lock_t lock){assert(lock==SD_LOCK_N64);return ++lock_calls==lock_fail?SD_ERROR_LOCKED:SD_OK;}
static void led_activity_on(void){on++;}
static void led_activity_off(void){off++;}
static void hw_adc_read_voltage_temperature(uint16_t *v,int16_t *t){adc_calls++;*v=3300;*t=25;}
static void sd_start_read(uint32_t address,uint32_t count){assert(address == DATA_BUFFER_ADDRESS && count >= 1 && count <= READ_AT_MAX_BLOCKS);read_address=address;read_count=count;starts++;}
static void sd_abort(void){aborts++;}
static int sd_cmd(unsigned cmd,uint32_t sector,unsigned rsp,void *unused){
 (void)unused;
 if(cmd==12){assert(rsp==RSP_R1b);cmd12++;return 0;}
 assert(cmd==18 && rsp==RSP_R1);cmd18++;
 if(fault==1)return 1;
 for(unsigned i=0;i<read_count*SD_SECTOR_SIZE;i++)bram[i]=(unsigned char)(sector+(sdstate.byte_swap?(i^1):i));
 return 0;
}
static dat_status_t sd_sync(unsigned timeout){assert(timeout==TIMEOUT_DATA_MS);return fault==2?DAT_ERROR_IO:fault==3?DAT_ERROR_TIMEOUT:DAT_OK;}
#include "read-at-sd.inc"
#include PHOS_READ_AT_OWNER
struct result {uint32_t cursor,r0,r1,status;unsigned locks,on,off,starts,aborts,cmd18,cmd12;unsigned char data[DATA_BUFFER_SIZE];};
static void reset(unsigned lf,unsigned fail,unsigned init,unsigned swap,unsigned block){
 memset(&p,0,sizeof(p));p.sd_card_sector=0xCAFE0123;memset(bram,0xA5,sizeof(bram));
 sdstate.card_initialized=init;sdstate.byte_swap=swap;sdstate.card_type_block=block;
 lock_fail=lf;fault=fail;lock_calls=on=off=starts=aborts=cmd18=cmd12=adc_calls=0;reply0=reply1=status=read_address=read_count=0;
}
static struct result snapshot(void){struct result r={0};r.cursor=p.sd_card_sector;r.r0=reply0;r.r1=reply1;r.status=status;r.locks=lock_calls;r.on=on;r.off=off;r.starts=starts;r.aborts=aborts;r.cmd18=cmd18;r.cmd12=cmd12;memcpy(r.data,bram,sizeof(bram));return r;}
int main(void){
 assert(CMD_ID_SD_SECTOR_SET=='I' && CMD_ID_SD_READ=='s' && CMD_ID_SD_READ_AT=='R' && CMD_ID_DIAGNOSTIC_GET=='%');
 assert(DIAGNOSTIC_ID_READ_AT==0x50485241);
 const uint32_t lbas[]={0,1234,0xFFFFFFF0,0xFFFFFFFF};
 const uint32_t counts[]={0,1,2,15,16,17,31,256,0x10000,0x800000,0xFFFFFFFF};unsigned cases=0;
 for(unsigned l=0;l<4;l++)for(unsigned n=0;n<11;n++)for(unsigned lf=0;lf<3;lf++)
 for(unsigned init=0;init<2;init++)for(unsigned swap=0;swap<2;swap++)for(unsigned block=0;block<2;block++)for(unsigned fail=0;fail<4;fail++){
  uint32_t count=counts[n],lba=lbas[l];
  reset(lf,fail,init,swap,block);p.data[0]=lba;legacy_command('I');
  if(!status){p.data[0]=SC64_BUFFERS_BASE;p.data[1]=count;legacy_command('s');}
  struct result a=snapshot();
  reset(lf,fail,init,swap,block);p.data[0]=lba;p.data[1]=count;read_at_command('R');
  struct result b=snapshot();assert(!memcmp(&a,&b,sizeof(a)));assert(on==off);
  if(lf==1){assert(p.sd_card_sector==0xCAFE0123 && reply0==(2u<<24|SD_ERROR_LOCKED));}
  else if(count==0 || (count>READ_AT_MAX_BLOCKS && count<0x800000)){assert(p.sd_card_sector==lba && reply0==(2u<<24|SD_ERROR_INVALID_ADDRESS) && !starts);}
  else if(count>=0x800000){assert(p.sd_card_sector==lba && reply0==(2u<<24|SD_ERROR_INVALID_ARGUMENT) && !starts);}
  else if(status)assert(p.sd_card_sector==lba);
  else assert(p.sd_card_sector==lba+count);
  cases++;
 }
 const uint32_t overflow_counts[]={0x700100,0x780000,0x7FFFFE,0x7FFFFF};
 for(unsigned i=0;i<4;i++)for(unsigned lf=0;lf<3;lf++)for(unsigned init=0;init<2;init++){
  reset(lf,0,init,0,1);p.data[0]=1234;p.data[1]=overflow_counts[i];read_at_command('R');
  assert(status && !starts && !on && !off);
  assert(reply0==(2u<<24|(lf==1?SD_ERROR_LOCKED:SD_ERROR_INVALID_ADDRESS)));
  assert(p.sd_card_sector==(lf==1?0xCAFE0123:1234));
 }
 reset(0,0,1,0,1);p.data[0]=0xFFFFFFFF;p.data[1]=0;read_at_command('%');assert(status && reply0==0x01000004 && !lock_calls && !starts && !adc_calls);
 reset(0,0,1,0,1);p.data[0]=DIAGNOSTIC_ID_READ_AT;p.data[1]=0;read_at_command('%');assert(!status && reply0==0x52415431 && reply1==0x52010010 && !lock_calls && !starts && !adc_calls && p.sd_card_sector==0xCAFE0123);
 reset(0,0,1,0,1);p.data[0]=0;read_at_command('%');assert(!status && reply1==(3300u<<16|25) && adc_calls==1);
 assert(cases==4224);
 puts("PASS READ_AT MCU: 4224 legacy differential + 24 overflow cases, diagnostics and side effects");
}

File: tests/read_at.py
#!/usr/bin/env python3
"""Test current MCU READ_AT against current legacy I/s with real source constants.

Generated owner extracts and artifacts stay in the explicit output directory.
The protocol model supplies only lock, DMA, command, LED and ADC boundaries;
production CFG translation/dispatch/diagnostics and SD reads execute unchanged.
"""
import argparse
import hashlib
import json
import os
from pathlib import Path
import re
import resource
import shlex
import subprocess

TESTS = Path(__file__).resolve().parent

def function(source, name):
    matches = list(re.finditer(r'^(?:static\s+)?\w+\s+' + re.escape(name) + r'\s*\([^;]*?\)\s*\{', source, re.M))
    if len(matches) != 1:
        raise RuntimeError('Missing or duplicate actual function: ' + name)
    start = matches[0].start()
    end = source.index('{', start) + 1
    depth = 1
    while depth:
        depth += (source[end] == '{') - (source[end] == '}')
        end += 1
    return source[start:end] + '\n'

def declaration(source, kind, name):
    values = [m[0] for m in re.finditer(r'typedef\s+' + kind + r'\s*\{[^}]*\}\s*\w+\s*;', source, re.S)
              if re.search(r'\}\s*' + name + r'\s*;$', m[0])]
    if len(values) != 1:
        raise RuntimeError('Missing or duplicate actual declaration: ' + name)
    return values[0] + '\n'

def macro(source, name):
    values = re.findall(r'^#define[ \t]+' + name + r'[ \t]+[^\n]+', source, re.M)
    if len(values) != 1:
        raise RuntimeError('Missing or duplicate actual macro: ' + name)
    return values[0] + '\n'

def section(source, first, last):
    if source.count(first) != 1 or source.count(last) != 1:
        raise RuntimeError('Missing or duplicate command boundary: ' + first)
    return source[source.index(first):source.index(last)]

def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--source', type=Path, default=TESTS.parent)
    parser.add_argument('--output', type=Path, required=True)
    args = parser.parse_args()
    root = args.source.resolve()
    out = args.output.resolve()
    protected = [TESTS, root / 'tests', root / 'sw']
    if out == root or any(out == path or path in out.parents for path in protected):
        raise RuntimeError('Generated output must be outside production source and tests/')
    out.mkdir(parents=True, exist_ok=True)
    resource.setrlimit(resource.RLIMIT_CORE, (0, 0))
    paths = ['sw/controller/src/cfg.c', 'sw/controller/src/sd.c',
             'sw/controller/src/sd.h', 'sw/bootloader/src/sc64.h']
    sources = {name: (root / name).read_text(encoding='utf-8') for name in paths}
    cfg, sd, header, n64 = (sources[name] for name in paths)
    types = ''.join(declaration(cfg, 'enum', name) for name in
                    ['cmd_id_t', 'diagnostic_id_t', 'translate_type_t', 'error_type_t', 'cfg_error_t'])
    types += ''.join(macro(cfg, name) for name in ['DATA_BUFFER_ADDRESS', 'DATA_BUFFER_SIZE'])
    types += ''.join(declaration(sd, 'enum', name) for name in ['rsp_type_t', 'dat_status_t'])
    types += ''.join(macro(sd, name) for name in ['TIMEOUT_DATA_MS', 'DAT_BLOCK_MAX_COUNT'])
    types += declaration(n64, 'struct', 'sc64_buffers_t') + macro(n64, 'SC64_BUFFERS_BASE')
    process = function(cfg, 'cfg_process')
    first = '        case CMD_ID_SD_SECTOR_SET:'
    read_at = '        case CMD_ID_SD_READ_AT:'
    read = '        case CMD_ID_SD_READ:'
    write = '        case CMD_ID_SD_WRITE:'
    selected = section(process, first, write)
    atomic = section(process, read_at, read)
    legacy = selected.replace(atomic, '')
    if selected.count(atomic) != 1:
        raise RuntimeError('READ_AT case extraction changed')
    diagnostic = process[process.index('        case CMD_ID_DIAGNOSTIC_GET:'):]
    prefix = function(cfg, 'cfg_translate_address') + function(cfg, 'cfg_read_diagnostic_data')
    def commands(cases, name):
        return 'static void ' + name + '(unsigned cmd) { switch (cmd) {\n' + cases + diagnostic
    owners = prefix + commands(legacy, 'legacy_command') + commands(selected, 'read_at_command')
    cursor = '            p.sd_card_sector = p.data[0];'
    if atomic.count(cursor) != 1:
        raise RuntimeError('Cursor negative mutation boundary changed')
    wrong_cursor = atomic.replace(cursor, '            (void)p.data[0];')
    bound = '            if (p.data[1] > DATA_BUFFER_SIZE / SD_SECTOR_SIZE && p.data[1] < 0x800000) {\n                return cfg_cmd_reply_error(ERROR_TYPE_SD_CARD, SD_ERROR_INVALID_ADDRESS);\n            }\n'
    if atomic.count(bound) != 1:
        raise RuntimeError('Bounds negative mutation boundary changed')
    variants = {
        'read-at-owner.inc': owners,
        'negative-cursor.inc': prefix + commands(legacy, 'legacy_command') + commands(selected.replace(atomic, wrong_cursor), 'read_at_command'),
        'negative-bound.inc': prefix + commands(legacy, 'legacy_command') + commands(selected.replace(atomic, atomic.replace(bound, '')), 'read_at_command'),
        'read-at-types.inc': types,
        'read-at-sd.inc': re.sub(r'\bp\.', 'sdstate.', function(sd, 'sd_read_sectors')),
        'sd.h': header,
        'read_at.c': (TESTS / 'read_at.c').read_text(encoding='utf-8'),
    }
    for name, text in variants.items():
        with (out / name).open('w', encoding='utf-8', newline='\n') as stream:
            stream.write(text)
    hashes = {name: hashlib.sha256(text.encode('utf-8')).hexdigest() for name, text in sources.items()}
    hashes['tests/read_at.c'] = hashlib.sha256((TESTS / 'read_at.c').read_bytes()).hexdigest()
    hashes['tests/read_at.py'] = hashlib.sha256(Path(__file__).read_bytes()).hexdigest()
    (out / 'sources.json').write_text(json.dumps(hashes, indent=2) + '\n', encoding='utf-8')
    env = dict(os.environ, ASAN_OPTIONS='halt_on_error=1:detect_leaks=1', UBSAN_OPTIONS='halt_on_error=1')
    log = out / 'results.log'
    log.write_text('', encoding='utf-8')
    def command(argv, negative=None):
        result = subprocess.run(argv, cwd=out, env=env, universal_newlines=True, stdout=subprocess.PIPE,
                                stderr=subprocess.STDOUT, timeout=120)
        with log.open('a', encoding='utf-8') as stream:
            stream.write('$ ' + ' '.join(shlex.quote(arg) for arg in argv) + '\n' + result.stdout)
        if negative:
            if not result.returncode or negative not in result.stdout:
                raise RuntimeError('Missing specific expected failure: ' + negative + '\n' + result.stdout)
        elif result.returncode:
            raise RuntimeError(result.stdout)
        return result.stdout
    flags = ['-std=c11', '-O1', '-g', '-Wall', '-Wextra', '-Werror', '-fsanitize=address,undefined',
             '-fno-sanitize-recover=all', '-fno-omit-frame-pointer']
    for variant, extract, negative in [
        ('read-at', 'read-at-owner.inc', None),
        ('negative-cursor', 'negative-cursor.inc', 'Assertion `!memcmp(&a,&b,sizeof(a))'),
        ('negative-bound', 'negative-bound.inc', 'Assertion `status && !starts && !on && !off')]:
        command(shlex.split(os.environ.get('CC', 'cc')) + flags +
                ['-DPHOS_READ_AT_OWNER="' + extract + '"', 'read_at.c', '-o', variant])
        output = command([str(out / variant)], negative)
        if negative:
            print('PASS rejected ' + variant + ' at expected assertion')
        else:
            if 'PASS READ_AT MCU: 4224 legacy differential + 24 overflow cases' not in output:
                raise RuntimeError('Missing complete fixture result')
            print(output, end='')
    print('PASS current-source READ_AT MCU gate; no hardware tested')

if __name__ == '__main__':
    main()
